### PR TITLE
fix #4708 refining what is possible via the requestconfig

### LINKED ISSUE #4708: Refine what is possible via withRequestConfig
The RequestConfig object exposes quite a few timeouts, authentication properties, and other things that we are allowing to be set on a per request basis if one starts with NamespacedKubernetesClient.withRequestConfig.

There are several issues with this:
- in the code it's effectively part of the Config, which is otherwise immutable.  That is confusing to deal with and requires extra logic to pass around to the interceptors.
- passing alternative authentication is confusing to the logic in the interceptors.  Although that may be improved by #4702.

It would be good to re-assess what about this feature is actually needed, and if it should just be part of the regular dsl - that is instead of client.adapt(NamespacedKubernetesClient.class).withRequestConfig(new RequestConfigBuilder().withImpersonateUsername("name").build()).(c -> ...)

should we just offer

client.withImpersonateUsername("name")...

With methods for whatever seems still useful.

### PR DESCRIPTION
**Pull request by shawkins**
## Description

Fixes #4708

There are several goals here.  The first is to ensure that only config values that can be changed without creating a new httpclient can be set on the RequestConfig - that means connectionTimeout cannot be there.  The next is that allowing for authentication related overrides complicates the logic and any documentation we'd want to create around the refresh interceptors, so it's not a good idea to allow that to be set on a per request basis.  It also seems unnecessary to override the oauthtokenprovider as the user already has control over how the token should be provided.

The next goal was to minimize our changes to the Config (the full application of those changes is often unclear and/or may require a new httpclient) - after this the config is nearly immutable and not copied - we're no longer creating copies to override the RequestConfig, nor for the namespace handling.  This means that places in the client which were referencing the config to get RequestConfig values were changed to reference a contextual RequestConfig.  The only logic that is still modifying the config is the token refresh - and this will likely get refined with #4802.  

To clean up how Interceptors reference the config, there is not an explicit parameter that finds tags associated with the request - this is not actually on the HttpRequest as we create those in other places and it's simpler not to have to find the tags that were applicable at that point and only deal with this concept as part of the Interceptors.  Due to this and the change to using the standard logic we can pass the full request, not just the headers into the interceptor methods.  After #4825 we'll be able to also add a generalized after method for regular requests - this should give us the full set of methods needed to introduce generalized logging interception.

This did not attempt to introduce new entry logic for override the requestconfig - it is still through the withRequestConfig method.  If we do provide other DSL entry points for such modifications they should use the same internal method for constructing a new Client / derived HttpClient that will have the appropriate overrides.

This does have conflicts with #4825 - but they will be straight-forward to resolve after one of them is committed.

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
Tried to quickly merge master but it needs more work :sweat:

**Comment by shawkins** [comment]
> Tried to quickly merge master but it needs more work

That's likely - there have been some overlapping changes of late.  After this goes in and #4877, then the extra fields will be removed as well.

**Comment by manusa** [comment]
Shall we merge #4877 first? (seems cleaner at the moment)

**Comment by shawkins** [comment]
> Shall we merge https://github.com/fabric8io/kubernetes-client/pull/4877 first? (seems cleaner at the moment)

That's good with me, I'll clean this one up after that.

**Comment by shawkins** [comment]
@manusa should be good now

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/http/Interceptor.java [hunk 1]
```diff
@@ -16,41 +16,23 @@
...
-   *
-   * @param config
-   * @return
-   */
-  default Interceptor withConfig(Config config) {
-    return this;
-  }
+  public interface RequestTags {
+
+    <T> T getTag(Class<T> type);
```
Comment (manusa) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/http/Interceptor.java:25]: In the future we should document what a Request tag is and what can it be used for

**Review by manusa — APPROVED** [review]

**Pull request merged**